### PR TITLE
Refactor documentation for `gensim.models.phrases`

### PR DESCRIPTION
**Pull request by CLearERR**
@menshikh-iv

### CODE REVIEW

File: gensim/models/phrases.py [hunk 1]
```diff
@@ -129,6 +136,22 @@ def _is_single(obj):
 class SentenceAnalyzer(object):
 
     def score_item(self, worda, wordb, components, scorer):
+        """Get sentence statistics.
+
+        Parameters
+        ----------
+        worda : str
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: don't forget about descriptions

File: gensim/models/phrases.py [hunk 2]
```diff
@@ -190,12 +217,17 @@ class PhrasesTransformation(interfaces.TransformationABC):
...
-            older Phrases/Phraser versions which did not support  pluggable scoring functions.
-            Otherwise, relies on utils.load
-        """
+        """Load a previously saved Phrases/Phraser class. Handles backwards compatibility from older Phrases/Phraser
+        versions which did not support pluggable scoring functions. Otherwise, relies on utils.load.
 
+        Parameters
+        ----------
+        args : object
+            Sequence of arguments, see :meth:`...` for more information.
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: `...`? you should link to `SaveLoad.load` I think

File: gensim/models/phrases.py [hunk 3]
```diff
@@ -226,47 +258,39 @@ def load(cls, *args, **kwargs):
...
 class Phrases(SentenceAnalyzer, PhrasesTransformation):
-    """
-    Detect phrases, based on collected collocation counts. Adjacent words that appear
-    together more frequently than expected are joined together with the `_` character.
-
-    It can be used to generate phrases on the fly, using the `phrases[sentence]`
-    and `phrases[corpus]` syntax.
-
+    """Detect phrases, based on collected collocation counts. Adjacent words that appear together more frequently than
+    expected are joined together with the `_` character. It can be used to generate phrases on the fly,
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: `_` - this can be changed

File: gensim/models/phrases.py [hunk 4]
```diff
@@ -226,47 +258,39 @@ def load(cls, *args, **kwargs):
...
-        on how much available memory you have.
-
-        `delimiter` is the glue character used to join collocation tokens, and
-        should be a byte string (e.g. b'_').
-
-        `scoring` specifies how potential phrases are scored for comparison to the `threshold`
-        setting. `scoring` can be set with either a string that refers to a built-in scoring function,
-        or with a function with the expected parameter names. Two built-in scoring functions are available
-        by setting `scoring` to a string:
+        sentences : list of str, optional
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: `iterable of list of str`

File: gensim/models/phrases.py [hunk 5]
```diff
@@ -226,47 +258,39 @@ def load(cls, *args, **kwargs):
...
+            The `sentences` iterable can be simply a list, but for larger corpora,
+            consider a generator that streams the sentences directly from disk/network,
+            without storing everything in RAM. See :class:`~gensim.models.word2vec.BrownCorpus`,
+            :class:`~gensim.models.word2vec.Text8Corpus` or :class:`~gensim.models.word2vec.LineSentence`
+            in the :mod:`~gensim.models.word2vec` module for such examples.
+        min_count : int, optional
+            Ignore all words and bigrams with total collected count lower
+            than this.
+        threshold : int, optional
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: float

File: gensim/models/phrases.py [hunk 6]
```diff
@@ -226,47 +258,39 @@ def load(cls, *args, **kwargs):
...
+            Represent a score threshold for forming the phrases (higher means fewer phrases).
+            A phrase of words `a` followed by `b` is accepted if the score of the
+            phrase is greater than threshold. See the `scoring` setting.
+        max_vocab_size : int, optional
+            Maximum size of the vocabulary. Used to control pruning of less common words, to keep memory under control.
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: better to add what's value we limit (this is number of tokens, not megabytes)

File: gensim/models/phrases.py [hunk 7]
```diff
@@ -226,47 +258,39 @@ def load(cls, *args, **kwargs):
...
+            The default of 40M needs about 3.6GB of RAM; increase/decrease `max_vocab_size` depending on how much
+            available memory you have.
+        delimiter : str, optional
+            Glue character used to join collocation tokens, should be a byte string (e.g. b'_').
+        scoring : {'default', 'npmi'} http://www.sphinx-doc.org/en/master/rest.html
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: what's a link?

File: gensim/models/phrases.py [hunk 8]
```diff
@@ -226,47 +258,39 @@ def load(cls, *args, **kwargs):
...
+            Specify how potential phrases are scored for comparison to the `threshold` setting.
+            `scoring` can be set with either a string that refers to a built-in scoring function, or with a function
+            with the expected parameter names. Two built-in scoring functions are available by setting `scoring` to a
+            string:
 
         'default': from "Efficient Estimaton of Word Representations in Vector Space" by
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: missing part (don't forget to use enumerate list)

File: gensim/models/phrases.py [hunk 9]
```diff
@@ -352,9 +380,16 @@ def __init__(self, sentences=None, min_count=5, threshold=10.0,
...
-        """
-        Load a previously saved Phrases class. Handles backwards compatibility from
-            older Phrases versions which did not support  pluggable scoring functions.
+        """Load a previously saved Phrases class. Handles backwards compatibility from older Phrases versions
+        which did not support pluggable scoring functions.
+
+        Parameters
+        ----------
+        args : object
+            Sequence of arguments, see :meth:`...` for more information.
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: same comments as for previous `load`

File: gensim/models/phrases.py [hunk 10]
```diff
@@ -373,7 +408,17 @@ def __str__(self):
     @staticmethod
     def learn_vocab(sentences, max_vocab_size, delimiter=b'_', progress_per=10000,
                     common_terms=frozenset()):
-        """Collect unigram/bigram counts from the `sentences` iterable."""
+        """Collect unigram/bigram counts from the `sentences` iterable. #TODO: Через пустой Phrasers
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: `#TODO: Через пустой Phrasers` - so Russian :D

File: gensim/models/phrases.py [hunk 11]
```diff
@@ -592,23 +649,26 @@ def pseudocorpus(self, phrases_model):
...
+        """Score, retained from original dataset."""
         try:
             return self.phrasegrams[tuple(components)][1]
         except KeyError:
             return -1
 
     def __getitem__(self, sentence):
-        """
-        Convert the input tokens `sentence` (=list of unicode strings) into phrase
+        """Convert the input tokens `sentence` (=list of unicode strings) into phrase
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: don't use this `(=list of unicode strings)`, better to write concrete types for arguments.

File: gensim/models/phrases.py [hunk 12]
```diff
@@ -68,11 +70,6 @@
 >>> print(bigram[sent])
 [u'the', u'mayor', u'shows', u'his', u'lack_of_interest']
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: You should fix this example

File: gensim/models/phrases.py [hunk 13]
```diff
@@ -226,58 +260,45 @@ def load(cls, *args, **kwargs):
...
+            available memory you have.
+        delimiter : str, optional
+            Glue character used to join collocation tokens, should be a byte string (e.g. b'_').
+        scoring : {'default', 'npmi'}, optional
+            Specify how potential phrases are scored for comparison to the `threshold` setting.
+            `scoring` can be set with either a string that refers to a built-in scoring function, or with a function
+            with the expected parameter names. Two built-in scoring functions are available by setting `scoring` to a
+            string:
+
+            1. `default` - :meth:`~gensim.models.phrases.original_scorer`.
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: this is `:func:`, not `:meth:`

File: gensim/models/phrases.py [hunk 14]
```diff
@@ -295,8 +316,23 @@ def __init__(self, sentences=None, min_count=5, threshold=10.0,
...
+        using the `phrases[sentence]` and `phrases[corpus]` syntax.
+
+        Example
+        ----------
+        >>> from gensim.test.utils import datapath
+        >>> from gensim.models.word2vec import Text8Corpus
+        >>> from gensim.models.phrases import Phrases
+        >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> bigram = Phrases(sentences, min_count=5, threshold=100)
+        >>> print bigram
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: That's a bad example, you should to show bigram extraction here

File: gensim/models/phrases.py [hunk 15]
```diff
@@ -352,9 +388,16 @@ def __init__(self, sentences=None, min_count=5, threshold=10.0,
...
-        """
-        Load a previously saved Phrases class. Handles backwards compatibility from
-            older Phrases versions which did not support  pluggable scoring functions.
+        """Load a previously saved Phrases class. Handles backwards compatibility from older Phrases versions
+        which did not support pluggable scoring functions.
+
+        Parameters
+        ----------
+        args : object
+            Sequence of arguments, see :class:`~gensim.models.phrases.Phrases` for more information.
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: Incorrect links, you should refer to parent class (i.e. `SaveLoad.load`)

File: gensim/models/phrases.py [hunk 16]
```diff
@@ -373,7 +416,37 @@ def __str__(self):
...
+            total number of words.
+
+        Example
+        ----------
+        >>> from gensim.test.utils import datapath
+        >>> from gensim.models.word2vec import Text8Corpus
+        >>> from gensim.models.phrases import Phrases
+        >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> learned = Phrases.learn_vocab(sentences,40000)
+        >>> print learned
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: what is it?

File: gensim/models/phrases.py [hunk 17]
```diff
@@ -411,8 +484,20 @@ def learn_vocab(sentences, max_vocab_size, delimiter=b'_', progress_per=10000,
...
+        sentences : list of str
+            List of unicode strings.
+
+        Example
+        -------
+        >>> from gensim.test.utils import datapath
+        >>> from gensim.models.word2vec import Text8Corpus
+        >>> from gensim.models.phrases import Phrases
+        >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> bigram = Phrases(sentences, min_count=5, threshold=100)
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: and where is `add_vocab`?

File: gensim/models/phrases.py [hunk 18]
```diff
@@ -473,22 +570,27 @@ def export_phrases(self, sentences, out_delimiter=b' ', as_tuples=False):
...
+            List of unicode strings.
+
+        Example
+        ----------
+        >>> from gensim.test.utils import datapath
+        >>> from gensim.models.word2vec import Text8Corpus
+        >>> from gensim.models.phrases import Phrases
+        >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> bigram = Phrases(sentences, min_count=5, threshold=100)
+        >>> bigram["trees","grass"]
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: list of .., i.e. should be `bigram[["trees", "grass"]]`, also, this have sense to make example that produced bigram

File: gensim/models/phrases.py [hunk 19]
```diff
@@ -527,19 +629,73 @@ def __getitem__(self, sentence):
 
 # calculation of score based on original mikolov word2vec paper
 def original_scorer(worda_count, wordb_count, bigram_count, len_vocab, min_count, corpus_word_count):
+    """
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: Where is a description of function (the same question for other scorer)

File: gensim/models/phrases.py [hunk 20]
```diff
@@ -527,19 +629,73 @@ def __getitem__(self, sentence):
...
+        Second word for comparison.
+    components : TODO
+    scorer : {'default', 'npmi'}
+        Scorer function, as given to :class:`~gensim.models.phrases.Phrases`.
+
+    Notes
+    -----
+    from "Efficient Estimaton of Word Representations in Vector Space" by
+    Mikolov, et. al.:
+    :math:`(count(worda, wordb) - min_count) * N /
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: did you check that this is rendered correctly?

File: gensim/models/phrases.py [hunk 21]
```diff
@@ -527,19 +629,73 @@ def __getitem__(self, sentence):
...
+        First word for comparison.
+    wordb : str
+        Second word for comparison.
+    components : TODO
+    scorer : {'default', 'npmi'}
+        Scorer function, as given to :class:`~gensim.models.phrases.Phrases`.
+
+    Notes
+    -----
+    from "Efficient Estimaton of Word Representations in Vector Space" by
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: when you refer a paper - please add it as hyperlink (always).

File: gensim/models/phrases.py (hunk 20 above)
Comment (menshikh-iv) [inline gensim/models/phrases.py]: also, this is a good idea to move it to description of function

File: gensim/models/phrases.py [hunk 22]
```diff
@@ -569,6 +726,22 @@ class Phraser(SentenceAnalyzer, PhrasesTransformation):
...
+            Phrases class object.
+
+        Example
+        -------
+        >>> from gensim.test.utils import datapath
+        >>> from gensim.models.word2vec import Text8Corpus
+        >>> from gensim.models.phrases import Phrases, Phraser
+        >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> phrases_model = Phrases(sentences, min_count=5, threshold=100)
+        >>> phraser_model = Phraser(phrases_model)
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: and what? how to use this classes?

File: gensim/models/phrases.py [hunk 23]
```diff
@@ -588,27 +761,103 @@ def __init__(self, phrases_model):
...
+        >>> pseudo = phraser_model.pseudocorpus(phrases_model)
+        //>>> phraser_model.score_item("tree","human",pseudo,'default')
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: ???

File: gensim/models/phrases.py [hunk 24]
```diff
@@ -588,27 +761,103 @@ def __init__(self, phrases_model):
...
+        >>> phraser_model.score_item(u"tree",u"human",pseudo,'default')
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: I'm not sure about demonstration of this function, this isn't really needed

File: gensim/models/phrases.py [hunk 25]
```diff
@@ -588,27 +761,103 @@ def __init__(self, phrases_model):
...
 
+        Example
+        -------
+        >>> from gensim.test.utils import datapath
+        >>> from gensim.models.word2vec import Text8Corpus
+        >>> from gensim.models.phrases import Phrases, Phraser
+        >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> phrases_model = Phrases(sentences, min_count=5, threshold=100)
+        >>> phraser_model = Phraser(phrases_model)
+        >>> pseudo = phraser_model.pseudocorpus(phrases_model)
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: why did this need?

File: gensim/models/phrases.py [hunk 26]
```diff
@@ -588,27 +761,103 @@ def __init__(self, phrases_model):
...
+        >>> phraser_model["tree", "human"]
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: incorrect, please use `phraser_model[["tree", "human"]]`

**Comment by menshikh-iv** [comment]
Good work @CLearERR :+1:

File: gensim/models/phrases.py [hunk 27]
```diff
@@ -933,12 +956,40 @@ def score_item(self, worda, wordb, components, scorer):
         >>> from gensim.models.word2vec import Text8Corpus
         >>> from gensim.models.phrases import Phrases, Phraser
         >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
+        >>> #train the detector with
```
Comment (piskvorky) [inline gensim/models/phrases.py]: PEP8: `#` followed by one ` ` space (here and elsewhere).

File: gensim/models/phrases.py [hunk 28]
```diff
@@ -978,6 +1029,44 @@ def __getitem__(self, sentence):
...
+        >>> #Train the detector with:
+        >>> phrases = Phrases(sentences, min_count=1, threshold=1)
+        >>> # Create a Phraser object to transform any sentence and turn 2 suitable tokens into 1 phrase:
+        >>> phraser_model = Phraser(phrases)
+        >>> #Input is a corpus:
+        >>> sent = [[u'trees', u'graph', u'minors'],[u'graph', u'minors']]
+        >>> #So we get 2 phrases
+        >>> res = phraser_model[sent]
+        >>> for phrase in res:
+        >>>     print phrase
```
Comment (piskvorky) [inline gensim/models/phrases.py]: Best use brackets, for py3k compatibility.

**Pull request merged**